packages/select/src/components/select/select.tsx

```diff
@@ -185,19 +191,32 @@ export class Select<T> extends React.PureComponent<ISelectProps<T>, ISelectState
                 </div>
                 <div onKeyDown={handleKeyDown} onKeyUp={handleKeyUp}>
                     {filterable ? input : undefined}
-                    <Menu ulRef={listProps.itemsParentRef}>{this.renderItems(listProps)}</Menu>
+                    {this.renderDropdown(listProps)}
                 </div>
             </Popover>
         );
     };
 
-    private renderItems({ items, renderItem }: IQueryListRendererProps<T>) {
-        const { initialContent, noResults } = this.props;
-        if (initialContent != null && this.isQueryEmpty()) {
-            return initialContent;
+    private renderDropdown(listProps: IQueryListRendererProps<T>) {
+        const { dropdownRenderer, initialContent, noResults } = this.props;
+        const { items, itemsParentRef, renderItem } = listProps;
+        const maybeInitialContent = initialContent != null && this.isQueryEmpty() ? initialContent : null;
+        if (dropdownRenderer !== undefined) {
+            return dropdownRenderer({
+                initialContent: maybeInitialContent,
+                items,
+                itemsParentRef,
+                noResults,
+                renderItem,
+            });
+        } else {
+            const renderedItems = items.map(renderItem).filter(item => item != null);
+            return (
+                <Menu ulRef={itemsParentRef}>
+                    {maybeInitialContent || (renderedItems.length > 0 ? renderedItems : noResults)}
+                </Menu>
```

Comment:
how about we pull this `else` block out to a static function and make it the default prop value?? then this conditional happens automagically!

Reply:
Sorry, I don't follow. Could you provide the function signature and point out where we'd call it?

Reply:
`static defaultProps: { menuRenderer: defaultMenuRenderer }` basically. does that make sense?
if not, no worries. i can make that refactor later.

Reply:
Ah yes, I see what you mean. Looks like you left a comment about how generic type variables can't be used in `defaultProps`, so let's use the same approach here as in `renderQueryList`:

https://github.com/palantir/blueprint/blob/2403cab2e93e2ea15b4f7cf2cfe126019f2426c8/packages/select/src/components/select/select.tsx#L145-L147